python/cuda_parallel/tests/test_iterators.py

```diff
@@ -0,0 +1,58 @@
+from cuda.parallel.experimental.iterators import (
+    CacheModifiedInputIterator,
+    ConstantIterator,
+    CountingIterator,
+    TransformIterator,
+)
+import cupy as cp
+import numpy as np
+
+
+def test_constant_iterator_equality():
+    assert ConstantIterator(np.int32(0)) == ConstantIterator(np.int32(0))
+    assert ConstantIterator(np.int32(0)) != ConstantIterator(np.int32(1))
+    assert ConstantIterator(np.int32(0)) != ConstantIterator(np.int64(0))
+
+
+def test_counting_iterator_equality():
+    assert CountingIterator(np.int32(0)) == CountingIterator(np.int32(0))
+    assert CountingIterator(np.int32(0)) != CountingIterator(np.int32(1))
+    assert CountingIterator(np.int32(0)) != CountingIterator(np.int64(0))
+
+
+def test_cache_modified_input_iterator_equality():
+    ary1 = cp.asarray([0, 1, 2])
+    ary2 = cp.asarray([3, 4, 5])
+    assert CacheModifiedInputIterator(ary1, "stream") == CacheModifiedInputIterator(
+        ary1, "stream"
+    )
+    assert CacheModifiedInputIterator(ary1, "stream") != CacheModifiedInputIterator(
+        ary2, "stream"
+    )
+
+
+def test_equality_transform_iterator():
+    def op1(x):
+        return x
+
+    def op2(x):
+        return 2 * x
+
+    def op3(x):
+        return x
+
+    it = CountingIterator(np.int32(0))
+    assert TransformIterator(it, op1) == TransformIterator(it, op1)
+    assert TransformIterator(it, op1) != TransformIterator(it, op2)
+    assert TransformIterator(it, op1) == TransformIterator(it, op3)
+
+    ary1 = cp.asarray([0, 1, 2])
+    ary2 = cp.asarray([3, 4, 5])
+    assert TransformIterator(ary1, op1) == TransformIterator(ary1, op1)
+    assert TransformIterator(ary1, op1) != TransformIterator(ary1, op2)
+    assert TransformIterator(ary1, op1) == TransformIterator(ary1, op3)
+    assert TransformIterator(ary1, op1) != TransformIterator(ary2, op1)
```

Comment:
I suppose along similar lines, can a `build_reduce` object built for `CountingIterator(np.int32(0))` be reused for `CountingIterator(np.int32(1))`? I suppose yes...

Reply:
I like [other the discussion](https://github.com/NVIDIA/cccl/pull/2788#discussion_r1873235103) we had on another PR since it leaves little room for doubts here. We should just follow what happens in C++.

Runtime comparison of two iterators of the same type should yield false if the *state* of these iterators is different:
```c++
struct op_t
{
  __host__ __device__ int operator()(int x) const { return x * 2; }
};

int main()
{
  int data[] = {1, 2};
  auto t1    = thrust::make_transform_iterator(data, op_t{});
  auto t2    = thrust::make_transform_iterator(data + 1, op_t{});

  if (t1 != t2)
  {
    std::cout << "not equal" << std::endl;
  }
}
```

Regarding the build step, it makes sense to match the logic of C++ as well. When you have `cub::DeviceReduce::Reduce(..., t1, ...)` and `cub::DeviceReduce::Reduce(..., t2, ...)` the implicit instantiation leads to only one instance of reduction, because types match. So what should go into the build cache is types of iterators rather than iterators themselves.

Reply:
OK, thanks. We need some similar notion of "iterator type" in Python then, because `type(it)` doesn't encode the same information as it does in C++. Thinking of a solution...